### PR TITLE
feat(web): plugin playground/enable edit html with modal

### PR DESCRIPTION
**Pull request by devshun**
# Overview

I implemented a modal for editing HTML code.
When editing HTML in the modal and submitting, the changes are reflected in the code editor.

## Release Notes

- **New Features**
	- Introduced `HtmlEditModal` for editing HTML code within a modal interface.
	- Added functionality to manage YAML and JavaScript files for widget visualization.
	- Enhanced `Code` component with an HTML editor and improved state management.
	- Added utility functions for HTML manipulation and file type identification.
	- Updated plugin management to include demo widgets and source code updates.

- **Improvements**
	- Enhanced `Viewer` component to accept dynamic widgets for better integration with the visualizer.
	- Streamlined hooks for improved code execution and file management in the Plugin Playground.

### CODE REVIEW

File: web/src/beta/features/PluginPlayground/Code/index.tsx [hunk 1]
```diff
@@ -9,37 +16,54 @@ type Props = {
...
-  }
-};
-
 const Code: FC<Props> = ({
   fileTitle,
   sourceCode,
   onChangeSourceCode,
   executeCode
 }) => {
+  const ediableHtmlSourceCode = useMemo(
```
Comment (mulengawilfred) [inline web/src/beta/features/PluginPlayground/Code/index.tsx:25]: Sorry, I think there is a small typo here. Should it be `editableHtmlSourceCode` ?

File: web/src/beta/features/PluginPlayground/Code/index.tsx [hunk 2]
```diff
@@ -9,37 +16,54 @@ type Props = {
...
-      />
-    </Wrapper>
+    <>
+      <Wrapper>
+        <Header>
+          <Button icon="playRight" iconButton onClick={executeCode} />
+          <Button
+            icon="pencilSimple"
+            title="HTML Editor"
+            disabled={ediableHtmlSourceCode === undefined}
```
Comment (mulengawilfred) [inline web/src/beta/features/PluginPlayground/Code/index.tsx:47]: Maybe we could have this as `disabled={!editableHtmlSourceCode}`. Let me know your thoughts.

**Review by mulengawilfred — APPROVED** [review]
Thanks @devshun ! This is great! I've approved as the functionality works as expected. Just two small comments added.

File: web/src/beta/features/PluginPlayground/Code/utils.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+const REEARTH_HTML_INJECTION_PATTERN =
+  /reearth\.(ui|modal|popup)\.show\(\s*`([^]*?)`\s*\);/g;
```
Comment (airslice) [inline web/src/beta/features/PluginPlayground/Code/utils.ts:2]: For these APIs they don't require the content be wrapped by ``, it also possible wrapped by '' or "";
And we also need to consider there could be mulitple API calls exists, the ideal UX is when user's corsor is located within one of the HTML fregment, enable the button, and when user click the button display the edit modal with the certain HTML fregment. You can treat this as next step if don't have good way to achieve that right now.

File: web/src/beta/features/PluginPlayground/Code/utils.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+};
+
+export const injectHtmlIntoSourceCode = (
+  htmlSourceCode: string,
+  sourceCode: string
+) => {
+  return sourceCode.replace(
+    REEARTH_HTML_INJECTION_PATTERN,
+    (_, type) => `reearth.${type}.show(\`${htmlSourceCode}\`);`
+  );
```
Comment (airslice) [inline web/src/beta/features/PluginPlayground/Code/utils.ts:16]: Same as above, there could be mulitple modal.show popup.show ui.show exist within one source code.

File: web/src/beta/features/PluginPlayground/Code/index.tsx (hunk 1 above)
Comment (devshun) [reply web/src/beta/features/PluginPlayground/Code/index.tsx]: thanks!
I've fixed this commit
https://github.com/reearth/reearth-visualizer/pull/1236/commits/2e27ee6475a631c58c6b198c30ea89d66c16be7b

File: web/src/beta/features/PluginPlayground/Code/index.tsx (hunk 2 above)
Comment (devshun) [reply web/src/beta/features/PluginPlayground/Code/index.tsx]: Thank you. I originally wrote it this way with the strict intent of only targeting undefined. However, after considering it further, I realized that enabling disabled for empty strings as well would be appropriate, so I’ve revised it as you suggested.

https://github.com/reearth/reearth-visualizer/pull/1236/commits/8c39c0d0b42595a06a579e9bccbad1f3c84e728c

File: web/src/beta/features/PluginPlayground/Code/utils.ts (hunk 3 above)
Comment (devshun) [reply web/src/beta/features/PluginPlayground/Code/utils.ts]: Thank you!
As you pointed out, I’ve added support for backticks and single quotes.
https://github.com/reearth/reearth-visualizer/pull/1236/commits/d3cae8778d9e9c985bdf0aaa4df440d6a57d154a
Considering "multiple reearth.show calls" might take some time to design, so I’ll handle that in a separate pull request.

File: web/src/beta/features/PluginPlayground/Code/hook.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                    {
+                      id: cur.id,
+                      __REEARTH_SOURCECODE: file.sourceCode
+                    }
+                  ]
+                }
+              }
+            }
+          }
+        };
```
[the remark this answers is not in the document]
Comment (mulengawilfred) [reply web/src/beta/features/PluginPlayground/Code/hook.ts:128]: Sorry. Let's consider this change if possible.

**Review by mulengawilfred — COMMENTED** [review]
Hey @devshun . I've added a small comment and I think there is a ci build failed. Once those are checked I can approve. Looks good!

**Review by airslice — APPROVED** [review]

**Pull request merged**